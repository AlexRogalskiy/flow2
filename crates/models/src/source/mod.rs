use crate::tables;

mod loader;
pub mod scenarios;
mod scope;
pub mod specs;
pub mod wrappers;

pub use loader::{FetchResult, LoadError, Loader};
pub use scope::Scope;

pub use specs::*;
pub use wrappers::*;

#[derive(Default, Debug)]
pub struct Tables {
    pub captures: tables::Captures,
    pub collections: tables::Collections,
    pub derivations: tables::Derivations,
    pub endpoints: tables::Endpoints,
    pub errors: tables::Errors,
    pub fetches: tables::Fetches,
    pub imports: tables::Imports,
    pub materializations: tables::Materializations,
    pub nodejs_dependencies: tables::NodeJSDependencies,
    pub projections: tables::Projections,
    pub resources: tables::Resources,
    pub schema_docs: tables::SchemaDocs,
    pub test_steps: tables::TestSteps,
    pub transforms: tables::Transforms,
}

impl Tables {
    // Access Tables as an array of dynamic TableObj instances.
    pub fn as_tables(&self) -> Vec<&dyn tables::TableObj> {
        // This de-structure ensures we can't fail to update if fields change.
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }

    // Access Tables as an array of mutable dynamic TableObj instances.
    pub fn as_tables_mut(&mut self) -> Vec<&mut dyn tables::TableObj> {
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }
}
