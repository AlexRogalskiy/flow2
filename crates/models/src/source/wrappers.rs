use schemars::{schema, JsonSchema};
use serde::{Deserialize, Serialize};
use serde_json::{from_value as from_json_value, json};

pub use protocol::flow::shuffle::Hash as ShuffleHash;

/// EndpointType enumerates the endpoint types understood by Flow.
#[derive(Copy, Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub enum EndpointType {
    Postgres,
    Sqlite,
    S3,
    GS,
    Remote,
}

impl EndpointType {
    pub fn as_scheme(&self) -> &str {
        match self {
            Self::Postgres => "postgres",
            Self::Sqlite => "sqlite",
            Self::S3 => "s3",
            Self::GS => "gs",
            Self::Remote => "remote",
        }
    }
}

/// ContentType enumerates resource content types understood by Flow.
#[derive(Copy, Debug, Clone, Serialize, Deserialize)]
pub enum ContentType {
    CatalogSpec,
    JsonSchema,
    NpmPack,
}

/// Names consist of Unicode letters, numbers, and symbols: - _ . /
///
/// Spaces and other special characters are disallowed.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialOrd, PartialEq, Ord, Eq, Hash)]
#[schemars(example = "CollectionName::example")]
pub struct CollectionName(#[schemars(schema_with = "CollectionName::schema")] String);

impl CollectionName {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn example() -> Self {
        Self("a/collection".to_owned())
    }
    fn schema(_: &mut schemars::gen::SchemaGenerator) -> schema::Schema {
        from_json_value(json!({
            "type": "string",
            "pattern": "^[^ \t\n\\!@#$%^&*()+=\\<\\>?;:'\"\\[\\]\\|~`]+$",
        }))
        .unwrap()
    }
}

impl std::ops::Deref for CollectionName {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialOrd, PartialEq, Ord, Eq)]
#[schemars(example = "TransformName::example")]
pub struct TransformName(String);

impl TransformName {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn example() -> Self {
        Self("a transform".to_owned())
    }
}

impl std::ops::Deref for TransformName {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// EndpointName names a Flow endpoint.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialOrd, PartialEq, Ord, Eq)]
#[schemars(example = "TransformName::example")]
pub struct EndpointName(String);

impl EndpointName {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn example() -> Self {
        Self("an endpoint".to_owned())
    }
}

impl std::ops::Deref for EndpointName {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// MaterializationName names a Flow materialization.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialOrd, PartialEq, Ord, Eq)]
#[schemars(example = "TransformName::example")]
pub struct MaterializationName(String);

impl MaterializationName {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn example() -> Self {
        Self("a materialization".to_owned())
    }
}

impl std::ops::Deref for MaterializationName {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// CaptureName names a Flow capture.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialOrd, PartialEq, Ord, Eq)]
#[schemars(example = "TransformName::example")]
pub struct CaptureName(String);

impl CaptureName {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn example() -> Self {
        Self("a capture".to_owned())
    }
}

impl std::ops::Deref for CaptureName {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// TestName names a Flow catalog test.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialOrd, PartialEq, Ord, Eq)]
#[schemars(example = "TransformName::example")]
pub struct TestName(String);

impl TestName {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn example() -> Self {
        Self("a capture".to_owned())
    }
}

impl std::ops::Deref for TestName {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// A URL identifying a resource, which may be a relative local path
/// with respect to the current resource (i.e, ../path/to/flow.yaml),
/// or may be an external absolute URL (i.e., http://example/flow.yaml).
#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[schemars(example = "RelativeUrl::example_relative")]
#[schemars(example = "RelativeUrl::example_absolute")]
pub struct RelativeUrl(String);

impl RelativeUrl {
    pub fn example_relative() -> Self {
        Self("../path/to/local.yaml".to_owned())
    }
    pub fn example_absolute() -> Self {
        Self("https://example/resource".to_owned())
    }
}

impl std::ops::Deref for RelativeUrl {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// JSON Pointer which identifies a location in a document.
#[derive(Serialize, Deserialize, Debug, JsonSchema, Clone, PartialEq, Eq)]
#[schemars(example = "JsonPointer::example")]
pub struct JsonPointer(#[schemars(schema_with = "JsonPointer::schema")] String);

impl JsonPointer {
    pub fn new(ptr: impl Into<String>) -> Self {
        Self(ptr.into())
    }
    pub fn example() -> Self {
        Self("/json/ptr".to_owned())
    }
    fn schema(_: &mut schemars::gen::SchemaGenerator) -> schema::Schema {
        from_json_value(json!({
            "type": "string",
            "pattern": "^/.+",
        }))
        .unwrap()
    }
}

impl std::ops::Deref for JsonPointer {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

impl AsRef<str> for JsonPointer {
    fn as_ref(&self) -> &str {
        &self.0
    }
}
