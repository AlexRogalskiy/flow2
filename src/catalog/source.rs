use super::{sql_params, Collection, ContentType, Error, Resource, Result, DB};
use crate::specs::build as specs;
use url::Url;

/// Source represents a top-level catalog build input.
#[derive(PartialEq, Eq, Debug, Copy, Clone)]
pub struct Source {
    pub resource: Resource,
}

impl Source {
    /// Register an Estuary Source specification with the catalog.
    pub fn register(db: &DB, uri: Url) -> Result<Source> {
        let source = Source {
            resource: Resource::register(db, ContentType::CatalogSpec, &uri)?,
        };
        if source.resource.is_processed(db)? {
            return Ok(source);
        }
        source.resource.mark_as_processed(db)?;

        let spec = source.resource.content(db)?;
        let spec: specs::Catalog = serde_yaml::from_slice(&spec)?;

        for uri in &spec.import {
            let uri = source.resource.join(db, uri)?;
            let import = Self::register(db, uri.clone()).map_err(|err| Error::At {
                loc: format!("import {:?}", uri),
                detail: Box::new(err),
            })?;
            Resource::register_import(db, source.resource, import.resource)?;
        }

        for (package, version) in spec.node_dependencies.iter() {
            db.prepare_cached("INSERT INTO nodejs_dependencies (package, version) VALUES (?, ?);")?
                .execute(sql_params![package, version])?;
        }

        for spec in &spec.collections {
            Collection::register(db, source, spec).map_err(|err| Error::At {
                loc: format!("collection {:?}", spec.name),
                detail: Box::new(err),
            })?;
        }
        Ok(source)
    }
}

#[cfg(test)]
mod test {
    use super::{
        super::{dump_tables, init_db_schema, open, Source},
        *,
    };
    use rusqlite::params as sql_params;
    use serde_json::json;

    #[test]
    fn test_register() -> Result<()> {
        let db = open(":memory:")?;
        init_db_schema(&db)?;

        let fixture = json!({
            "import": [
                "../other/spec",
                "test://example/other/spec",
            ],
            "nodeDependencies": {
                "package-one": "v0.1.2",
                "pkg-2": "~v2",
            },
            "collections": [
                {
                    "name": "a/collection",
                    "schema": "test://example/schema",
                    "key": ["/key"],
                },
            ],
        });
        db.execute(
            "INSERT INTO resources (resource_id, content_type, content, is_processed) VALUES
                    (1, 'application/vnd.estuary.dev-catalog-spec+yaml', CAST(? AS BLOB), FALSE),
                    (2, 'application/vnd.estuary.dev-catalog-spec+yaml', CAST('{}' AS BLOB), FALSE),
                    (10, 'application/schema+yaml', CAST('true' AS BLOB), FALSE);",
            sql_params![fixture],
        )?;
        db.execute(
            "INSERT INTO resource_urls (resource_id, url, is_primary) VALUES
                    (1, 'test://example/main/spec', TRUE),
                    (2, 'test://example/other/spec', TRUE),
                    (10, 'test://example/schema', TRUE);",
            sql_params![],
        )?;
        Source::register(&db, Url::parse("test://example/main/spec")?)?;

        // Expect other catalog spec & schema were processed.
        assert!(Resource { id: 2 }.is_processed(&db)?);
        assert!(Resource { id: 10 }.is_processed(&db)?);

        assert_eq!(
            dump_tables(
                &db,
                &["resource_imports", "collections", "nodejs_dependencies"]
            )?,
            json!({
                "resource_imports": [[1, 2], [1, 10]],
                "collections": [[1, "a/collection", "test://example/schema", ["/key"], 1]],
                "nodejs_dependencies": [["package-one", "v0.1.2"], ["pkg-2", "~v2"]],
            }),
        );

        Ok(())
    }
}
