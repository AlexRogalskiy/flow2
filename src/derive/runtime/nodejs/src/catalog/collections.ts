// This file contains placeholder examples of collection types,
// which are over-written during catalog build.

type ExampleSourceCollection = {
  hello: string;
};

type ExampleDerivedCollection = {
  world: string;
};
