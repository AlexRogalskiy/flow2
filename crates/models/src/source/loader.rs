use super::wrappers::{CollectionName, ContentType, ShuffleHash, TransformName};
use super::{specs, Scope, Tables};

use doc::Schema as CompiledSchema;
use futures::future::{FutureExt, LocalBoxFuture};
use json::schema::{build::build_schema, Application, Keyword};
use std::cell::RefCell;
use std::future::Future;
use url::Url;

#[derive(thiserror::Error, Debug)]
pub enum LoadError {
    #[error("failed to parse URL")]
    URLParse(#[from] url::ParseError),
    #[error("failed to fetch resource {uri}")]
    Fetch {
        uri: String,
        #[source]
        detail: Box<dyn std::error::Error + Send + Sync>,
    },
    #[error("failed to parse YAML (location {:?})", .0.location())]
    YAMLErr(#[from] serde_yaml::Error),
    #[error("failed to merge YAML alias nodes")]
    YAMLMergeErr(#[from] yaml_merge_keys::MergeKeyError),
    #[error("failed to build JSON schema")]
    SchemaBuild(#[from] json::schema::BuildError),
    #[error("failed to index JSON schema")]
    SchemaIndex(#[from] json::schema::index::Error),
}

pub type FetchResult = Result<Box<[u8]>, Box<dyn std::error::Error + Send + Sync>>;

/// Loader provides a stack-based driver for traversing catalog source
/// models, with dispatch to a Visitor trait and having fine-grained
/// tracking of location context.
pub struct Loader<F, FF>
where
    F: FnMut(&Url) -> FF,
    FF: Future<Output = FetchResult>,
{
    // Tables loaded by the build process.
    tables: RefCell<Tables>,
    // Dynamic fetch function for retrieving discovered, unvisited resources.
    fetch: RefCell<F>,
}

impl<F, FF> Loader<F, FF>
where
    F: FnMut(&Url) -> FF,
    FF: Future<Output = FetchResult>,
{
    /// Build and return a new Loader.
    pub fn new(tables: Tables, fetch: F) -> Loader<F, FF> {
        Loader {
            tables: RefCell::new(tables),
            fetch: RefCell::new(fetch),
        }
    }

    pub fn into_tables(self) -> Tables {
        self.tables.into_inner()
    }

    /// Load (or re-load) a resource of the given ContentType.
    pub async fn load_resource<'a>(
        &'a self,
        scope: Scope<'a>,
        resource: &'a Url,
        content_type: ContentType,
    ) {
        // Mark as visited, so that recursively-loaded imports don't re-visit.
        self.tables.borrow_mut().fetches.push_row(resource.clone());

        let content = (self.fetch.borrow_mut())(&resource);
        let content = content.await.map_err(|e| LoadError::Fetch {
            uri: resource.to_string(),
            detail: e,
        });

        if let Some(content) = self.fallible(scope, content) {
            self.load_resource_content(scope, resource, &content, content_type)
                .await;
        }
    }

    // Resources are loaded recursively, and Rust requires that recursive
    // async calls be made through a boxed future. Otherwise, the generated
    // state machine would have infinite size!
    fn load_resource_content<'a>(
        &'a self,
        scope: Scope<'a>,
        resource: &'a Url,
        content: &'a [u8],
        content_type: ContentType,
    ) -> LocalBoxFuture<'a, ()> {
        async move {
            self.tables.borrow_mut().resources.push_row(
                resource.clone(),
                content_type,
                content.to_vec(),
            );
            let scope = scope.push_resource(&resource);

            match content_type {
                ContentType::CatalogSpec => self.load_catalog(scope, content).await,
                ContentType::JsonSchema => self.load_schema_document(scope, content).await,
                _ => None,
            };
            ()
        }
        .boxed_local()
    }

    async fn load_schema_document<'s>(&'s self, scope: Scope<'s>, content: &[u8]) -> Option<()> {
        let dom: serde_json::Value = self.fallible(scope, serde_yaml::from_slice(&content))?;
        let doc: CompiledSchema =
            self.fallible(scope, build_schema(scope.resource().clone(), &dom))?;

        let mut index = doc::SchemaIndex::new();
        self.fallible(scope, index.add(&doc))?;

        self.load_schema_node(scope, &index, &doc).await;

        self.tables
            .borrow_mut()
            .schema_docs
            .push_row(scope.flatten(), dom);

        Some(())
    }

    fn load_schema_node<'s>(
        &'s self,
        scope: Scope<'s>,
        index: &'s doc::SchemaIndex<'s>,
        schema: &'s CompiledSchema,
    ) -> LocalBoxFuture<'s, ()> {
        // Build an iterator that returns a Future for each Application keyword.
        // That future in turn:
        // * Builds a recursive future A from it's contained schema, and
        // * Builds a future B which fetches a referenced external resource, if present.
        // Then futures A & B are concurrently joined.
        let it = schema.kw.iter().filter_map(move |kw| match kw {
            Keyword::Application(app, child) => Some(async move {
                // Add Application keywords to the Scope's Location.
                let location = app.push_keyword(&scope.location);
                let scope = Scope {
                    location: app.push_keyword_target(&location),
                    ..scope
                };

                // Map to an external URL that's not contained by this CompiledSchema.
                let uri = match app {
                    Application::Ref(uri) => {
                        // $ref applications often use #fragment suffixes which indicate
                        // a sub-schema of the base schema document to use.
                        let mut uri = uri.clone();
                        uri.set_fragment(None);

                        if index.fetch(&uri).is_none() {
                            Some(uri)
                        } else {
                            None
                        }
                    }
                    _ => None,
                };

                // Recursive call to walk the schema.
                let recurse = self.load_schema_node(scope, index, child);

                if let Some(uri) = uri {
                    // Concurrently fetch |uri| while continuing to walk the schema.
                    let ((), ()) = futures::join!(
                        recurse,
                        self.load_import(scope, &uri, ContentType::JsonSchema)
                    );
                } else {
                    let () = recurse.await;
                }
            }),
            _ => None,
        });

        // Join all futures of the iterator.
        futures::future::join_all(it)
            .map(|_: Vec<()>| ())
            .boxed_local()
    }

    /// Load a schema reference, which may be an inline schema.
    async fn load_schema_reference<'s>(
        &'s self,
        scope: Scope<'s>,
        schema: specs::Schema,
    ) -> Option<Url> {
        // If schema is a relative URL, then import it.
        if let specs::Schema::Url(import) = schema {
            let mut import = self.fallible(scope, scope.resource().join(import.as_ref()))?;

            // Temporarily strip schema fragment to import base document.
            let fragment = import.fragment().map(str::to_string);
            import.set_fragment(None);

            self.load_import(scope, &import, ContentType::JsonSchema)
                .await;

            import.set_fragment(fragment.as_deref());
            Some(import)
        } else {
            // Schema is in-line. Create a synthetic resource URL by extending the parent
            // with a `ptr` query parameter, encoding the json pointer path of the schema.
            let mut import = scope.resource().clone();
            import.set_query(Some(&format!("ptr={}", scope.location.url_escaped())));

            self.load_resource_content(
                scope,
                &import,
                &serde_json::to_vec(&schema).unwrap(),
                ContentType::JsonSchema,
            )
            .await;

            self.tables
                .borrow_mut()
                .imports
                .push_row(scope.flatten(), scope.resource(), &import);
            Some(import)
        }
    }

    // Load an import to another resource, recursively fetching if not yet visited.
    async fn load_import<'s>(
        &'s self,
        scope: Scope<'s>,
        import: &'s Url,
        content_type: ContentType,
    ) {
        // Recursively process the import if it's not already visited.
        if !self
            .tables
            .borrow_mut()
            .fetches
            .iter()
            .any(|f| f.resource == *import)
        {
            self.load_resource(scope, &import, content_type).await;
        }

        self.tables
            .borrow_mut()
            .imports
            .push_row(scope.flatten(), scope.resource(), import);
    }

    // Load a top-level catalog specification.
    async fn load_catalog<'s>(&'s self, scope: Scope<'s>, content: &[u8]) -> Option<()> {
        let dom: serde_yaml::Value = self.fallible(scope, serde_yaml::from_slice(&content))?;
        let dom: serde_yaml::Value =
            self.fallible(scope, yaml_merge_keys::merge_keys_serde(dom))?;

        let specs::Catalog {
            _schema,
            import,
            node_dependencies,
            collections,
            endpoints,
            materializations,
            captures,
            tests,
        } = self.fallible(scope, serde_yaml::from_value(dom))?;

        // Collect NodeJS dependencies.
        for (package, version) in node_dependencies {
            let scope = scope
                .push_prop("nodeDependencies")
                .push_prop(&package)
                .flatten();

            self.tables
                .borrow_mut()
                .nodejs_dependencies
                .push_row(scope, package, version);
        }

        // Task which loads all imports.
        let import = import.into_iter().enumerate().map(|(index, import)| {
            async move {
                let scope = scope.push_prop("import");
                let scope = scope.push_item(index);

                // Map from relative to absolute URL.
                if let Some(import) = self.fallible(scope, scope.resource().join(import.as_ref())) {
                    self.load_import(scope, &import, ContentType::CatalogSpec)
                        .await;
                }
            }
        });
        let import = futures::future::join_all(import);

        // Task which loads all collections.
        let collections = collections
            .into_iter()
            .map(|(name, collection)| async move {
                self.load_collection(
                    scope.push_prop("collections").push_prop(name.as_ref()),
                    &name,
                    collection,
                )
                .await;
            });
        let collections = futures::future::join_all(collections);

        // Collect endpoints.
        for (name, endpoint) in endpoints {
            let scope = scope
                .push_prop("endpoints")
                .push_prop(name.as_ref())
                .flatten();

            self.tables.borrow_mut().endpoints.push_row(
                scope,
                name,
                endpoint.endpoint_type(),
                endpoint.base_config(),
            );
        }

        // Collect captures.
        for (name, capture) in captures {
            let scope = scope
                .push_prop("captures")
                .push_prop(name.as_ref())
                .flatten();

            let (allow_push, endpoint, patch_config) = match capture.inner {
                specs::CaptureType::PushAPI(config) => (true, None, config),
                specs::CaptureType::Endpoint(specs::EndpointRef { name, config }) => {
                    (false, Some(name), config)
                }
            };

            self.tables.borrow_mut().captures.push_row(
                scope,
                name,
                &capture.target.name,
                allow_push,
                endpoint,
                serde_json::Value::Object(patch_config),
            );
        }

        // Collect materializations.
        for (name, materialization) in materializations {
            let scope = scope
                .push_prop("materializations")
                .push_prop(name.as_ref())
                .flatten();

            let specs::Materialization {
                source: specs::MaterializationSource { name: source },
                endpoint:
                    specs::EndpointRef {
                        name: endpoint,
                        config: patch_config,
                    },
                fields,
            } = materialization;

            self.tables.borrow_mut().materializations.push_row(
                scope,
                name,
                source,
                endpoint,
                serde_json::Value::Object(patch_config),
                fields,
            );
        }

        // Collect tests.
        for (name, steps) in tests {
            for (index, step) in steps.into_iter().enumerate() {
                let scope = scope
                    .push_prop("tests")
                    .push_prop(&name)
                    .push_item(index)
                    .flatten();
                let name = name.clone();

                self.tables
                    .borrow_mut()
                    .test_steps
                    .push_row(scope, name, index as u32, step);
            }
        }

        let (_, _): (Vec<()>, Vec<()>) = futures::join!(import, collections);
        Some(())
    }

    async fn load_collection<'s>(
        &'s self,
        scope: Scope<'s>,
        collection_name: &'s CollectionName,
        collection: specs::Collection,
    ) {
        let specs::Collection {
            schema,
            key,
            store:
                specs::EndpointRef {
                    name: store,
                    config: store_patch_config,
                },
            projections,
            derivation,
        } = collection;

        // Visit all collection projections.
        for (field, projection) in projections.iter() {
            let (location, partition) = match projection {
                specs::Projection::Pointer(location) => (location, false),
                specs::Projection::Object {
                    location,
                    partition,
                } => (location, *partition),
            };

            self.tables.borrow_mut().projections.push_row(
                scope.push_prop("projections").push_prop(field).flatten(),
                collection_name,
                field,
                location,
                partition,
                true, // User-provided.
            );
        }

        // Task which loads & maps collection schema => URL.
        // Recoverable failures project to Ok(None).
        let schema = self.load_schema_reference(scope.push_prop("schema"), schema);

        // If this collection is a derivation, concurrently
        // load the collection's schema and its derivation.
        let schema = match derivation {
            Some(derivation) => {
                let derivation = self.load_derivation(
                    scope.push_prop("derivation"),
                    collection_name,
                    derivation,
                );

                let (schema, ()) = futures::join!(schema, derivation);
                schema
            }
            None => schema.await,
        };

        if let Some(schema) = schema {
            self.tables.borrow_mut().collections.push_row(
                scope.flatten(),
                collection_name,
                schema,
                key,
                store,
                serde_json::Value::Object(store_patch_config),
            );
        }
    }

    async fn load_derivation<'s>(
        &'s self,
        scope: Scope<'s>,
        derivation_name: &'s CollectionName,
        derivation: specs::Derivation,
    ) {
        let specs::Derivation {
            register:
                specs::Register {
                    schema: register_schema,
                    initial: register_initial,
                },
            transform,
        } = derivation;

        // Task which loads & maps register schema => URL.
        let register_schema = async move {
            self.load_schema_reference(
                scope.push_prop("register").push_prop("schema"),
                register_schema,
            )
            .await
        };

        // Task which loads each derivation transform.
        let transforms = transform.into_iter().map(|(name, transform)| async move {
            self.load_transform(
                scope.push_prop("transform").push_prop(name.as_ref()),
                &name,
                derivation_name,
                transform,
            )
            .await
        });
        let transforms = futures::future::join_all(transforms);

        let (register_schema, _): (_, Vec<()>) = futures::join!(register_schema, transforms);

        if let Some(register_schema) = register_schema {
            self.tables.borrow_mut().derivations.push_row(
                scope.flatten(),
                derivation_name,
                register_schema,
                register_initial,
            );
        }
    }

    async fn load_transform<'s>(
        &'s self,
        scope: Scope<'s>,
        transform_name: &'s TransformName,
        derivation: &'s CollectionName,
        transform: specs::Transform,
    ) {
        let specs::Transform {
            source:
                specs::TransformSource {
                    name: source,
                    schema: source_schema,
                    partitions: source_partitions,
                },
            read_delay,
            priority,
            shuffle,
            update,
            publish,
        } = transform;

        let (shuffle_key, shuffle_lambda, shuffle_hash) = match shuffle {
            Some(specs::Shuffle::Key(key)) => (Some(key), None, ShuffleHash::None),
            Some(specs::Shuffle::MD5(key)) => (Some(key), None, ShuffleHash::Md5),
            Some(specs::Shuffle::Lambda(lambda)) => (None, Some(lambda), ShuffleHash::None),
            None => (None, None, ShuffleHash::None),
        };

        // Map optional source schema => URL.
        let source_schema = match source_schema {
            Some(url) => {
                self.load_schema_reference(scope.push_prop("source").push_prop("schema"), url)
                    .await
            }
            None => None,
        };

        self.tables.borrow_mut().transforms.push_row(
            scope.flatten(),
            transform_name,
            derivation,
            source,
            source_partitions,
            source_schema,
            shuffle_key,
            shuffle_lambda,
            shuffle_hash,
            read_delay.map(|d| d.as_secs() as u32),
            priority,
            update,
            publish,
        );
    }

    // Consume a result capable of producing a LoadError.
    // Pass through a Result::Ok<T> as Some<T>.
    // Or, record a Result::Err<T> and return None.
    fn fallible<'s, T, E>(&self, scope: Scope<'s>, r: Result<T, E>) -> Option<T>
    where
        E: Into<LoadError>,
    {
        match r {
            Ok(t) => Some(t),
            Err(err) => {
                self.tables
                    .borrow_mut()
                    .errors
                    .push_row(scope.flatten(), anyhow::anyhow!(err.into()));
                None
            }
        }
    }
}
