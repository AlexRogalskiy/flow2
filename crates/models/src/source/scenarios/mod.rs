use crate::source;

use futures::channel::oneshot;
use futures::future::FutureExt;
use serde_json::Value;
use std::cell::RefCell;
use std::collections::BTreeMap;
use std::task::Poll;
use url::Url;

macro_rules! file_tests {
    ($($name:ident,)*) => {
    $(
        #[test]
        fn $name() {
			let fixture = include_bytes!(concat!(stringify!($name), ".yaml"));
            let fixture: Value = serde_yaml::from_slice(fixture).unwrap();
			let tables = evaluate_fixtures(Default::default(), &fixture);
			insta::assert_debug_snapshot!(tables);
        }
    )*
    }
}

file_tests! {
    test_catalog_import_cycles,
    test_collections,
    test_derivations,
    test_schema_with_inline,
    test_schema_with_nested_ids,
    test_schema_with_references,
    test_simple_catalog,
    test_test_case,
    test_endpoints_captures_materializations,
}

pub fn evaluate_fixtures(catalog: source::Tables, fixture: &Value) -> source::Tables {
    let fixtures = match fixture {
        Value::Object(m) => m,
        _ => panic!("fixtures must be an object having resource properties"),
    };

    // Fetches holds started fetches since the last future poll.
    // Use an ordered map so that we signal one-shots in a stable order,
    // making snapshots reliable.
    let fetches: RefCell<BTreeMap<String, oneshot::Sender<source::FetchResult>>> =
        RefCell::new(BTreeMap::new());

    // Fetch function which queues oneshot futures for started fetches.
    let fetch = |url: &Url| {
        let (tx, rx) = oneshot::channel();
        if let Some(_) = fetches.borrow_mut().insert(url.to_string(), tx) {
            panic!("url {} has already been fetched", url);
        }
        rx.map(|r| r.unwrap())
    };
    let loader = source::Loader::new(catalog, fetch);
    let root = Url::parse("test://root").unwrap();

    // What's going on here? Glad you asked.
    //
    // loader.load_resource() is returning a Future, which under the covers
    // is a big, synchronously-invoked, and deterministic state machine that's
    // built with the compiler's assistance.
    //
    // By construction, we know that there's only one real "await" point of this
    // future which will cause it to return Poll::Pending: when it's called into the
    // |fetch| closure we gave it above, and it has no work remaining to do until
    // at least one of the Futures returned by |fetch| resolves.
    //
    // Loader is walking sources concurrently. It processes fetches as separate
    // internal (still synchronous & deterministic) tasks, so we *do* expect to
    // see multiple |fetch| calls made in between Poll::Pending poll results of
    // the future.
    //
    // So, the strategy is this: we _synchronously_ poll the future forward in a
    // loop (e.x., we're not using a runtime *at all*). Every time it returns
    // Poll::Pending, we assert that it's queued new calls to |fetch| during this
    // iteration, and we resolve each of those response futures. Eventually, it
    // returns Poll::Ready when it's fully walked the source fixture, and at that
    // point we unwrap and return the loaded Tables.
    //
    // Note that the use of BTreeMap above is significant: it means that we resolve
    // those call futures in a stable (sorted) order, and that's the order with which
    // the future will process resolved responses on it's next poll(). It makes the
    // whole mess fully deterministic.

    let mut fut = loader
        .load_resource(
            source::Scope::new(&root),
            &root,
            source::ContentType::CatalogSpec,
        )
        .boxed_local();

    let waker = futures::task::noop_waker();
    let mut ctx = std::task::Context::from_waker(&waker);

    loop {
        match fut.poll_unpin(&mut ctx) {
            Poll::Ready(()) => {
                std::mem::drop(fut);
                return loader.into_tables();
            }
            Poll::Pending if fetches.borrow().is_empty() => {
                // Note the future can return Pending *only because* it's blocked
                // waiting for one or more |fetch| fixtures above to resolve.
                panic!("future is pending, but started no fetches")
            }
            Poll::Pending => {
                for (url, tx) in fetches.borrow_mut().split_off("") {
                    match fixtures.get(&url) {
                        Some(value) => tx.send(Ok(value.to_string().as_bytes().into())),
                        None => tx.send(Err("fixture not found".into())),
                    }
                    .unwrap();
                }
            }
        }
    }
}
