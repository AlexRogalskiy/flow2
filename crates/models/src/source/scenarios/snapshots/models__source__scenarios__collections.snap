---
source: crates/models/src/source/scenarios/mod.rs
expression: tables
---
Tables {
    captures: [],
    collections: [
        Collection {
            scope: test://root#/collections/test~1collection,
            collection: test/collection,
            schema: test://root/schema.json#foobar,
            key: ["/key/1","/key/0"],
            store_endpoint: endpoint,
            store_patch_config: {},
        },
    ],
    derivations: [],
    endpoints: [],
    errors: [],
    fetches: [
        Fetch {
            resource: test://root,
        },
        Fetch {
            resource: test://root/schema.json,
        },
    ],
    imports: [
        Import {
            scope: test://root#/collections/test~1collection/schema,
            from_resource: test://root,
            to_resource: test://root/schema.json,
        },
    ],
    materializations: [],
    nodejs_dependencies: [],
    projections: [
        Projection {
            scope: test://root#/collections/test~1collection/projections/field_a,
            collection: test/collection,
            field: field_a,
            location: /a/a,
            partition: 1,
            user_provided: 1,
        },
        Projection {
            scope: test://root#/collections/test~1collection/projections/field_b,
            collection: test/collection,
            field: field_b,
            location: /b/b,
            partition: 0,
            user_provided: 1,
        },
    ],
    resources: [
        Resource {
            resource: test://root,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://root/schema.json,
            content_type: "JsonSchema",
            content: ".. binary ..",
        },
    ],
    schema_docs: [
        SchemaDoc {
            schema: test://root/schema.json,
            dom: {"$anchor":"foobar","properties":{"a":{"properties":{"a":{"type":"string"}},"type":"object"},"b":{"properties":{"b":{"type":"string"}},"type":"object"},"key":{"items":{"type":"string"},"minItems":2,"type":"array"}},"type":"object"},
        },
    ],
    test_steps: [],
    transforms: [],
}
